### DISCUSSION TITLE
Micropython File Access on Spike Prime

### DISCUSSION THREAD

**Discussion by Redstonetill** [pybricks/support#1792] [General]
I just thought about a possibility to access a single file (I know that the Spike Prime (the hub I am using for pybricks) doesnt support a filesystem). So my generell idea was that the firmware already has a way to read and write code to the drive in the hub (I couldn't figure out what type of storage the hub is using, just 1MB flash and another 32MB of something unkown) since it can save code even if the hub is power off (I am aware of the 512byte storage already supported by the pybricks firmware but since it isnt that much space another would be great). So if the firmware is already able to read and store files, I dont think, that it should be to hard to write a function in the firmware to read parts of a file previously hard coded into the firmware.
I would love to hear the take of someone other on this and if possible some information where the read and write code is being stored in the firmware since I would love to try it out :)

**Comment by dlech** [comment]
Technically, it isn't saving "files". Basically, there is a region of RAM that gets saved to flash memory at power off and restored at power on. 

Most of the relevant code is at https://github.com/pybricks/pybricks-micropython/blob/master/lib/pbio/sys/storage.c

Pybricks currently only uses <1MB of the 32MB external flash for this so that we don't mess up the file system from the official LEGO firmware. So there isn't much free space to work with (that's why `PBSYS_CONFIG_STORAGE_USER_DATA_SIZE` is only 512 bytes).

**Reply by abliss** [reply]
If I were okay with never being able to restore my Spike Prime Hub to the stock firmware, what would be the easiest way to make that 32MB available for general purpose use? Would I just change `PBSYS_CONFIG_STORAGE_USER_DATA_SIZE` to a bigger number?

**Reply by dlech** [reply]
It also has to fit in RAM, so I don't think you could make it crazy big.

**Reply by abliss** [reply]
Thanks. Is it at all possible to write a big blob to the flash memory such that I could then load parts of it into RAM at runtime?  In particular I am thinking of uploading an image of the FLL mat and then reading it to help compare its RGB values to my color sensor readings.

**Reply by gte829d** [reply]
> n image of the FLL mat and then reading it to help compare its RGB values to my color sensor readings.

We explored this idea before the season started but ran out of time.  I would highly suggest have your team prototype this idea in Matlab, Octave or regular Python.   Just that prototype work can take you very far in the Robot Design part of the judging session.    In essence using the pair of HSV values that two color sensors give you, can you localize your robot base on the map (position and orientation).   HSV color distance is tricky due to the cone shape of the HSV color map.   Translating to RGB is a good idea but we found the accuracy of the trig functions to be more than desirable on the hub.   There are 3 values you can use,  the Left color sensor,  the Right color sensor and the "color distance" between the two.   This third quantity gives you some robustness to lighting conditions.    But your lookup would be what two points on the map with the spatial separation of your actual color sensors minimizes the color distances of each color sensor with the map and does the color distance between the actual two points relatively close to your measurement.   In the simulation you can quantize your map down to something that fits the HW but you'll also get an idea of how well you can expect it to work.

**Reply by romw314** [reply]
> If I were okay with never being able to restore my Spike Prime Hub to the stock firmware

Maybe you'd be also able to restore the stock firmware:

1. Flash using vanilla micropython (skip the backup step, there's no reason to back up Pybricks firmware): https://github.com/micropython/micropython/tree/master/ports/stm32/boards/LEGO_HUB_NO6#readme (this creates an empty filesystem on the hub)
2. Restore the original firmware using Pybricks Code.
